File: aoc/aoc2022/day04/Day4.cs
public class Day4
{
    public static void Run()
    {
        using (StreamReader file = new StreamReader("day04/p.in"))
        {
            var cnt = 0;
            string ln;
            while ((ln = file.ReadLine()) != null)
            {
                string[] words = ln.Split(',');
                var p1f = int.Parse(words[0].Split('-')[0]);
                var p1t = int.Parse(words[0].Split('-')[1]);
                var p2f = int.Parse(words[1].Split('-')[0]);
                var p2t = int.Parse(words[1].Split('-')[1]);

                if (p1f >= p2f && p1f <= p2t)
                {
                    cnt++;
                }
                else if (p1t >= p2f && p1t <= p2t)
                {
                    cnt++;
                }
                else if (p1f < p2f && p1t > p2t)
                {
                    cnt++;
                }
            }

            file.Close();
            Console.WriteLine(cnt);
        }
    }
}

File: aoc/aoc2019/day04/Day4.fs
module Day4
open System
open System.IO

type Day4() =
    static member private IsAscending (s: string) =
        s |> Seq.pairwise |> Seq.forall (fun (a, b) -> a <= b)

    static member private HasDouble (s: string) =
        s
        |> Seq.windowed 4
        |> Seq.exists (fun arr -> 
            (arr.[1] = arr.[2] && arr.[0] <> arr[1] && arr[3] <> arr[1]))

    static member private StartDouble (s: string) =
        s[0] = s[1] && s[2] <> s[0]

    static member private EndDouble (s: string) =
        s[s.Length - 1] = s[s.Length - 2] && s[s.Length - 3] <> s[s.Length - 1]

    static member Run() =
        use file = new StreamReader("day04/p.in")
        while not file.EndOfStream do
            let ln = file.ReadLine()
            if not (String.IsNullOrWhiteSpace ln) then
                let parts = ln.Split("-")
                let lo, hi = int parts.[0], int parts.[1]
                let cnt =
                    { lo .. hi }
                    |> Seq.map (fun i -> string i)
                    |> Seq.filter (fun s ->
                        Day4.IsAscending s && (Day4.HasDouble s || Day4.StartDouble s || Day4.EndDouble s))
                    |> Seq.length

                Console.WriteLine(cnt)

